optuna/samplers/cmaes.py

```diff
@@ -0,0 +1,367 @@
+        library as the backend.
+
+    Args:
+
+        x0:
+            A dictionary of an initial parameter values for CMA-ES. By default, the mean of ``low``
+            and ``high`` for each distribution is used.
+
+        sigma0:
+            Initial standard deviation of CMA-ES. By default, ``sigma0`` is set to
+            ``min_range / 6``, where ``min_range`` denotes the minimum range of the distributions
+            in the search space.
+
+        seed:
+            A random seed for CMA-ES.
+
+        n_startup_trials:
+            The independent sampling is used instead of the CMA-ES algorithm until the given number
+            of trials finish in the same study.
+
+        independent_sampler:
+            A :class:`~optuna.samplers.BaseSampler` instance that is used for independent
+            sampling. The parameters not contained in the relative search space are sampled
+            by this sampler.
+            The search space for :class:`~optuna.samplers.CmaEsSampler` is determined by
+            :func:`~optuna.samplers.intersection_search_space()`.
+
+            If :obj:`None` is specified, :class:`~optuna.samplers.RandomSampler` is used
+            as the default.
+
+            .. seealso::
+                :class:`optuna.samplers` module provides built-in independent samplers
+                such as :class:`~optuna.samplers.RandomSampler` and
+                :class:`~optuna.samplers.TPESampler`.
+
+        warn_independent_sampling:
+            If this is :obj:`True`, a warning message is emitted when
+            the value of a parameter is sampled by using an independent sampler.
+
+            Note that the parameters of the first trial in a study are always sampled
+            via an independent sampler, so no warning messages are emitted in this case.
+    """
+
+    def __init__(
+        self,
+        x0: Optional[Dict[str, Any]] = None,
+        sigma0: Optional[float] = None,
+        n_startup_trials: int = 1,
+        independent_sampler: Optional[BaseSampler] = None,
+        warn_independent_sampling: bool = True,
+        seed: Optional[int] = None,
+    ) -> None:
+
+        self._x0 = x0
+        self._sigma0 = sigma0
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._n_startup_trials = n_startup_trials
+        self._warn_independent_sampling = warn_independent_sampling
+        self._logger = optuna.logging.get_logger(__name__)
+        self._cma_rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self, study: "optuna.Study", trial: "optuna.structs.FrozenTrial",
+    ) -> Dict[str, BaseDistribution]:
+
+        # Import here to avoid circular imports without breaking a backward compatibility.
+        from optuna.samplers import intersection_search_space
+
+        search_space = {}  # type: Dict[str, BaseDistribution]
+        for name, distribution in intersection_search_space(study).items():
```

Comment:
Not a strong opinion but you can pass `ordered_dict=True` to `intersection_search_space`. That way you might not have to pass around your `ordered_keys` list. Performance-wise, it shouldn't be an issue.

Reply:
> That way you might not have to pass around your ordered_keys list. 

`ordered_keys` is used at `sample_relative` method, not `infer_relative_searchspace`. So it requires to change the type of return value from `Dict[str, BaseDistribution]` to `Union[Dict[...], OrderedDict[...]]`.

I think we should separate a pull request. It may be better to add a todo comment for now.
What do you think?

Reply:
I put a todo comment at 4c74eeb.

Reply:
Sounds good. By the way, using string type type hints seems to work for `OrderedDict` https://stackoverflow.com/questions/41207128/how-do-i-specify-ordereddict-k-v-types-for-mypy-type-annotation.